Change error code for empty table name

elonazoulay opened the pull request:
Changing the error code from GENERIC_INTERNAL to SYNTAX_ERROR
for queries where the table name is empty.

raghavsethi commented on presto-spi/src/main/java/com/facebook/presto/spi/SchemaUtil.java:
```diff
@@ -25,7 +27,7 @@ static String checkNotEmpty(String value, String name)
             throw new NullPointerException(name + " is null");
```
Is there a reason we wouldn't fix this one also? Or is it not possible?

elonazoulay replied:
There is no reason and it's possible:)
I'll update shortly.

findepi commented on presto-spi/src/main/java/com/facebook/presto/spi/SchemaUtil.java:
```diff
@@ -25,7 +27,7 @@ static String checkNotEmpty(String value, String name)
         }
         if (value.isEmpty()) {
-            throw new IllegalArgumentException(name + " is empty");
+            throw new PrestoException(GENERIC_USER_ERROR, name + " is empty");
```
This is not the right place so handle this, as this method does not even know what `name` is supposed to represent. This is simply `requireNonNull + checkArgument`.
IMO you should do the check in the proper place, where you handle user input.

elonazoulay replied:
Will do, thanks @findepi !

elonazoulay replied:
Question: should this be a parsing error similar to when the catalog/schema/table/column name is null?
For now assuming yes.

elonazoulay commented:
@findepi , in the visitTable function it looks like the input is checked so I added the check there and reverted the SchemaUtil changes. Is this closer to what you mentioned?

raghavsethi commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -838,6 +838,9 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
 
             QualifiedObjectName name = createQualifiedObjectName(session, table, table.getName());
+            if (name.getObjectName().isEmpty()) {
+                throw new SemanticException(MISSING_TABLE, table, "Tablename is empty");
```
```suggestion
                throw new SemanticException(MISSING_TABLE, table, "Table name is empty");
```

raghavsethi commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -158,6 +158,43 @@ public void testParsingError()
+    @Test
+    public void testEmptyTableName()
+    {
+        //assertQuery("SELECT * FROM \" \"");
```
?

elonazoulay replied:
That should have been removed, thanks!

raghavsethi commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -158,6 +158,43 @@ public void testParsingError()
+    @Test
+    public void testEmptySchemaName()
+    {
+        assertQueryFails("SELECT * FROM \"\".foo", "schemaName is empty");
```
Where does this error come from?

elonazoulay replied:
From SchemaUtil. Looking at adding a check in the visitTable function similar to the check for table name.

findepi commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -158,6 +158,42 @@ public void testParsingError()
+    @Test
+    public void testNullCatalogName()
+    {
+        assertQueryFails("SELECT * FROM NULL.schema.foo", "line 1:15: mismatched input 'NULL'. Expecting: .*");
```
There is no need to run these tests per each connector. Add them somewhere else than ATQ.

raghavsethi replied:
Good point, missed that!

findepi commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java:
```diff
@@ -130,6 +130,18 @@ public void testResetSession()
+    @Test
+    public void testNullColumnName()
+    {
+        assertQueryFails("CREATE TABLE foo (col1 BIGINT, null BIGINT)", "line 1:32: mismatched input 'null'. Expecting: .*");
```
There is no need to run these tests per each connector.

elonazoulay commented:
FYI, I will move these tests to TestAnalyzer shortly.

elonazoulay commented:
The tests are moved to TestAnalyzer.
Only empty table name and schema name are tested since those are the only ones affected by this change.
